### PR TITLE
Keras Objectives and 0-based label

### PR DESCRIPTION
**Pull request by hkvision**
Fix https://github.com/intel-analytics/zoo/issues/81

### CODE REVIEW

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/objectives/ClassNLLCriterion.scala [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+ * `sparse_categorical_crossentropy` loss in keras.
+ *
+ * @param weights weights of each element of the input
+ * @param sizeAverage size average of batch
+ * @param logProbAsInput indicating whether to accept log-probabilities or probabilities as input.
+ *                   True means accepting log-probabilities as input.
+ * @param ev numeric operator
+ * @tparam T numeric type
+ */
+class ClassNLLCriterion[T: ClassTag](
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/objectives/ClassNLLCriterion.scala]: change to keras style name, instead of torch style name

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/objectives/NetCriterion.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+package com.intel.analytics.zoo.pipeline.api.keras.objectives
+
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractCriterion, Activity}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+import scala.reflect.ClassTag
+
+abstract class NetCriterion[A <: Activity: ClassTag, B <: Activity: ClassTag, T: ClassTag]
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/objectives/NetCriterion.scala]: change to Loss or LossFunction?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/objectives/ClassNLLCriterion.scala (hunk 1 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/objectives/ClassNLLCriterion.scala]: wondering if we use Keras-style name, do we need the arguments for original torch criterion (such as sizeAverage etc)? As Keras won't have parameters for a criterion...

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/objectives/NetCriterion.scala (hunk 2 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/objectives/NetCriterion.scala]: okay

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/objectives/ClassNLLCriterion.scala (hunk 1 above)
Comment (jason-dai) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/objectives/ClassNLLCriterion.scala]: we can keep the additional augments (with default values)

**Comment by hkvision** [comment]
Manually test for text classification example on both scala and python for zero-based label and predict classes.

**Comment by jason-dai** [comment]
I'm OK with predictClasses

**Comment by hkvision** [comment]
jenkins passed http://10.239.47.211:18888/job/ZOO-PR-Validation/738/

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/utils/KerasUtils.scala [hunk 3]
```diff
@@ -402,4 +403,22 @@ object KerasUtils {
...
+  /**
+   * classes: RDD of 1-based label.
+   * If zeroBasedLabel is true, convert to RDD of 0-based label.
+   * Otherwise, just return classes itself.
+   */
+  def toZeroBasedLabel(
+      zeroBasedLabel: Boolean = true,
+      classes: RDD[Int]): RDD[Int] = {
+    if (zeroBasedLabel) {
+      classes.mapPartitions { partition =>
```
Comment (yangw1234) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/utils/KerasUtils.scala]: why not just classes.map(_ + 1)

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/metrics/Accuracy.scala [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ *                       If false, labels start from 1.
+ */
+class Accuracy[T](
+    val zeroBasedLabel: Boolean = true)(implicit ev: TensorNumeric[T])
+  extends Top1Accuracy[T] {
+
+  override def apply(output: Activity, target: Activity):
+  ValidationResult = {
+    if (zeroBasedLabel) {
+      super.apply(output, target.asInstanceOf[Tensor[T]].clone().apply1(x => ev.plus(x, ev.one)))
```
Comment (yangw1234) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/metrics/Accuracy.scala]: target.toTensor[T].clone().add(1.0f) ?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/metrics/Accuracy.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+ *                       If false, labels start from 1.
+ */
+class Top5Accuracy[T](
+    val zeroBasedLabel: Boolean = true)(implicit ev: TensorNumeric[T])
+  extends BigDLTop5Accuracy[T] {
+
+  override def apply(output: Activity, target: Activity):
+  AccuracyResult = {
+    if (zeroBasedLabel) {
+      super.apply(output, target.asInstanceOf[Tensor[T]].clone().apply1(x => ev.plus(x, ev.one)))
```
Comment (yangw1234) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/metrics/Accuracy.scala]: same as above

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/objectives/LossFunction.scala [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ * The base class for Keras-style API objectives in Analytics Zoo.
+ *
+ * @tparam A Input data type.
+ * @tparam B Target data type.
+ * @tparam T Numeric type of parameter. Only support float/double now.
+ */
+abstract class LossFunction[A <: Activity: ClassTag, B <: Activity: ClassTag, T: ClassTag]
+(implicit ev: TensorNumeric[T]) extends AbstractCriterion[A, B, T] {
+
+  val loss: AbstractCriterion[A, B, T]
```
Comment (yangw1234) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/objectives/LossFunction.scala]: private ?

**Comment by hkvision** [comment]
jenkins failed due to the pr https://github.com/intel-analytics/analytics-zoo/pull/476
All other unit tests passed. http://10.239.47.211:18888/view/ZOO-PR/job/ZOO-PR-Validation/742/
Merge it first. Thanks for your review! @yangw1234
